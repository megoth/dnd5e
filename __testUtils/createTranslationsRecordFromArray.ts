export default function createTranslationsRecordFromArray(
  url,
  list: string[]
): Record<string, string> {
  return list.reduce((memo: Record<string, string>, key) => {
    // eslint-disable-next-line no-param-reassign
    memo[`${url}#${key}`] = key;
    return memo;
  }, {});
}
