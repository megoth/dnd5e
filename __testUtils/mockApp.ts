import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
