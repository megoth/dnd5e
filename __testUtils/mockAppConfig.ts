export default function mockAppConfig() {
  return {
    errorsUrl: "http://localhost:3000/data/errors.ttl",
    translationsUrl: "http://localhost:3000/data/translations.ttl",
  };
}
