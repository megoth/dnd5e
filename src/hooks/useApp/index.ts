import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
