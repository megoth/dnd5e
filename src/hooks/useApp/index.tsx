import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
