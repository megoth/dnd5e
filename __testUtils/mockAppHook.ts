import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
