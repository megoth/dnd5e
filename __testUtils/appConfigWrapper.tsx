import React, { ReactNode } from "react";
import AppConfigProvider from "../src/contexts/appConfig";
import mockAppConfig from "./mockAppConfig";

interface Props {
  children: ReactNode;
}

export default function AppConfigWrapper({ children }: Props) {
  const { errorsUrl, translationsUrl } = mockAppConfig();
  return (
    <AppConfigProvider errorsUrl={errorsUrl} translationsUrl={translationsUrl}>
      {children}
    </AppConfigProvider>
  );
}
