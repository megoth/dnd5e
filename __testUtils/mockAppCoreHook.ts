import mockApp from "./mockApp";
import { AppCoreHookResult } from "../src/hooks/useAppCore";

export default function mockAppCoreHook(
  mock: jest.Mock,
  response: Partial<AppCoreHookResult> = {
    app: mockApp(),
  }
) {
  return mock.mockReturnValue(response);
}
