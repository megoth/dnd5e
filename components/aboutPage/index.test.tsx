import React from "react";
import { render } from "@testing-library/react";
import AboutPage, { TESTID_ABOUT_PAGE_LANGUAGE_WARNING } from "./index";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import { mockNorwegianApp } from "../../__testUtils/mockApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("AboutPage", () => {
  it("renders", () => {
    mockAppHook(mockedAppHook);
    const { asFragment } = render(<AboutPage markdown="# test" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows a warning on non-english pages", () => {
    mockAppHook(mockedAppHook, mockNorwegianApp());
    const { asFragment, getByTestId } = render(<AboutPage markdown="# test" />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_ABOUT_PAGE_LANGUAGE_WARNING)).toBeDefined();
  });
});
