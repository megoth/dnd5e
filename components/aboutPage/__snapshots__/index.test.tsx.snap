// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutPage renders 1`] = `
<DocumentFragment>
  <header>
    <a
      class="main-container main-container--content flex"
      href="/"
    >
      <picture>
        <source
          media="(min-width: 640px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-128.png"
        />
        <source
          media="(min-width: 480px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-96.png"
        />
        <img
          alt="[Translation for https://example.com/global/translations.ttl#appLogoDescription does not exist]"
          src="/icons/app/icons8-dungeons-and-dragons-64.png"
        />
      </picture>
      <div
        class="font-serif text-2xl sm:text-3xl md:text-4xl self-center antialiased"
      >
        [Translation for https://example.com/global/translations.ttl#appName does not exist]
      </div>
    </a>
  </header>
  <main
    class="main-container main-container--content my-1"
  >
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <h1
        id="test"
      >
        test
      </h1>
    </div>
  </main>
  <footer
    class="main-container main-container--content md:flex mt-2"
  >
    <div
      class="md:flex-grow mb-2"
    >
      <a
        class="link mr-1 p-1"
        href="/about"
      >
        [Translation for https://example.com/global/translations.ttl#aboutTitle does not exist]
      </a>
      <a
        class="link mr-1 p-1"
        href="/faq"
      >
        [Translation for https://example.com/global/translations.ttl#faqShort does not exist]
      </a>
    </div>
    <div
      class="mb-2"
    />
  </footer>
</DocumentFragment>
`;

exports[`AboutPage shows a warning on non-english pages 1`] = `
<DocumentFragment>
  <header>
    <a
      class="main-container main-container--content flex"
      href="/"
    >
      <picture>
        <source
          media="(min-width: 640px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-128.png"
        />
        <source
          media="(min-width: 480px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-96.png"
        />
        <img
          alt="[Translation for https://example.com/global/translations.ttl#appLogoDescription does not exist]"
          src="/icons/app/icons8-dungeons-and-dragons-64.png"
        />
      </picture>
      <div
        class="font-serif text-2xl sm:text-3xl md:text-4xl self-center antialiased"
      >
        [Translation for https://example.com/global/translations.ttl#appName does not exist]
      </div>
    </a>
  </header>
  <main
    class="main-container main-container--content my-1"
  >
    <div
      class="bg-yellow-200 text-yellow-900 border-yellow-400 border px-2 py-1 rounded-sm mb-2"
      data-testid="about-page-language-warning"
    >
      [Translation for https://example.com/global/translations.ttl#onlyAvailableInEnglish does not exist]
    </div>
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <h1
        id="test"
      >
        test
      </h1>
    </div>
  </main>
  <footer
    class="main-container main-container--content md:flex mt-2"
  >
    <div
      class="md:flex-grow mb-2"
    >
      <a
        class="link mr-1 p-1"
        href="/about"
      >
        [Translation for https://example.com/global/translations.ttl#aboutTitle does not exist]
      </a>
      <a
        class="link mr-1 p-1"
        href="/faq"
      >
        [Translation for https://example.com/global/translations.ttl#faqShort does not exist]
      </a>
    </div>
    <div
      class="mb-2"
    >
      <span
        data-testid="locale-selector-language"
      >
        <a
          class="button"
          href="/?locale=%F0%9F%87%B3%F0%9F%87%B4"
        >
          ðŸ‡ºðŸ‡¸Â [Translation for https://example.com/global/translations.ttl#translateTo does not exist]
        </a>
      </span>
    </div>
  </footer>
</DocumentFragment>
`;
